### PR TITLE
Support Arrays Of Arrays

### PR DESCRIPTION
**Pull request by jpjsonic**
## **Type**
enhancement

___

## **Description**
- Added support for surface charts in Visivo projects, enabling complex data visualization with arrays of arrays.
- New surface trace configuration allows for detailed 3D data representation.
- Introduced a surface chart with customizable dimensions and title, enhancing dashboard visualization capabilities.

___

## **Changes walkthrough**
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>project.visivo.yml</strong><dd><code>Add Support for Surface Charts with Arrays of Arrays</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
test-projects/simple-database/project.visivo.yml

<li>Added a new surface trace configuration supporting arrays of arrays <br>for columns.<br> <li> Introduced a new surface chart with custom dimensions and a title.<br> <li> Updated dashboards to include the new surface chart.<br>

</details>
    

  </td>
  <td><a href="https://github.com/visivo-io/visivo/pull/74/files#diff-13c8f6a460dcde5f97876ce1d1961c51b3e08fc2e6967a0e532a01b3bd1fc55d">+34/-1</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> ✨ **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

File: test-projects/simple-database/project.visivo.yml [hunk 1]
```diff
@@ -80,9 +80,21 @@ traces:
       x_data: x
     props:
       type: indicator
-      value: column(x_data)[0]
+      value: column(x_data)[-1]
       delta:
         reference: column(x_data)[1]
+  - name: surface-trace
+    model: ref(test table local)
+    # columns:
```
Comment (tgsoverly) [inline test-projects/simple-database/project.visivo.yml:88]: You wan these comments?

File: test-projects/simple-database/project.visivo.yml [hunk 2]
```diff
@@ -212,7 +224,15 @@ charts:
   - name: indicator_chart
     traces:
       - ref(Indicator Trace)
-
+  - name: surface-chart
```
Comment (tgsoverly) [inline test-projects/simple-database/project.visivo.yml:227]: Neato.

File: viewer/src/models/Trace.js [hunk 3]
```diff
@@ -25,25 +25,32 @@ const convertDotKeysToNestedObject = (flatObject) => {
...
+    if (Array.isArray(obj)) {
+        obj.forEach((item, index) => replaceColumnRefWithData(item, data, obj, index));
+    } else if (typeof obj === "object" && obj !== null) {
+        for (const prop in obj) {
+            replaceColumnRefWithData(obj[prop], data, obj, prop);
+        }
+    } else if (typeof obj === 'string' && obj.match(COLUMN_REGEX)) {
+        const match = obj.match(COLUMN_REGEX);
+        const columnName = match[1];
+        if (match[3] !== undefined && match[4] !== undefined) {
```
Comment (tgsoverly) [inline viewer/src/models/Trace.js:38]: I was just thinking we should replace the "3" and "4" with a more clear name.

**Review by tgsoverly — APPROVED** [review]
Do we need any docs changes?

**Comment by jpjsonic** [comment]
> Do we need any docs changes?

That's a good call out. I'll add some docs to the TraceColumn object that shows how you can slice.

**Pull request merged**